site/app/controllers/MiscController.php

```diff
@@ -227,6 +235,33 @@ public function downloadCourseFile($dir, $path) {
         readfile($path);
     }
 
+    /**
+     * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/downloadTestCaseResult")
+     */
+    public function downloadTestCaseResult($gradeable_id, $version, $test_case, $file_name, $user_id) {
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        $graded_gradeable = $this->tryGetGradedGradeable($gradeable, $user_id, false);
+        if ($user_id !== $this->core->getUser()->getId()) {
+            if (!$this->core->getAccess()->canI("grading.electronic.grade", ["gradeable" => $gradeable, "graded_gradeable" => $graded_gradeable])) {
+                $this->core->addErrorMessage("You do not have peremission to download this file!");
+                return new RedirectResponse($this->core->buildCourseUrl(['gradeable', $gradeable_id]));
+            }
+        }
+        $autograde = $graded_gradeable->getAutoGradedGradeable()->getAutoGradedVersionInstance($version);
+
+        $name = "details/test" . sprintf("%02d", $test_case) . "/$file_name";
+        $files = $autograde->getResultsFiles();
+        $file_path = $files[$name]['path'];
```

Comment:
Need to account for `$name` not existing in `$files` for an invalid `$test_case` or `$file_name`. Should do an `isset` check that shows error message and redirects on that case.